File: Source/3 Data mining/Implementation - Jan/Java/src/itu/datamining/week5/kMean/KMeans.java
package itu.datamining.week5.kMean;
import itu.datamining.week5.data.Iris;
import itu.datamining.week5.kMedoid.ReassignedResponse;

import java.util.ArrayList;
import java.util.Hashtable;

public class KMeans {

	public static ArrayList<KMeanCluster> KMeansPartition(int k, ArrayList<Iris> data)
	{
		ArrayList<KMeanCluster> clusters = getInitialPartioningCluster(k, data);

		ReassignedResponse response = null;
		do{
			response = reassignDataToCluster(data, clusters);
			clusters = response.newCluster;
		} while(response.reassigned);

		return clusters;
	}

	public static ReassignedResponse reassignDataToCluster(ArrayList<Iris> data, ArrayList<KMeanCluster> clusters){

		boolean reassigned = false;
		Hashtable<KMeanCluster, KMeanCluster> newOldClusterIndex = new Hashtable<>();

		for(KMeanCluster cluster : clusters){
			KMeanCluster newCluster = new KMeanCluster();
			newCluster.clusterMean = cluster.calcClusterMean();
			newOldClusterIndex.put(newCluster, cluster);
		}

		ArrayList<KMeanCluster> newCluster = new ArrayList<>(newOldClusterIndex.keySet());

		for(Iris record : data){
			KMeanCluster minManhattenDistanceCluster = null;
			float minDistance = Float.MAX_VALUE;

			for(KMeanCluster cluster : newOldClusterIndex.keySet()){
				float distance = manhattenDistance(record, cluster.clusterMean);
				if(minDistance > distance){
					minManhattenDistanceCluster = cluster;
					minDistance = distance;
				}
			}
			minManhattenDistanceCluster.ClusterMembers.add(record);

			if(!reassigned && !newOldClusterIndex.get(minManhattenDistanceCluster).ClusterMembers.contains(record)){
				reassigned = true;
			}
		}

		return new ReassignedResponse(reassigned, newCluster);
	}

	public static float manhattenDistance(Iris iris1, Iris iris2){
		return Math.abs(iris1.Sepal_Length - iris2.Sepal_Length)
				+ Math.abs(iris1.Sepal_Width - iris2.Sepal_Width)
				+ Math.abs(iris1.Petal_Length - iris2.Petal_Length)
				+ Math.abs(iris1.Petal_Width - iris2.Petal_Width);
	}

	public static ArrayList<KMeanCluster> getInitialPartioningCluster(int k, ArrayList<Iris> data) {
		ArrayList<KMeanCluster> initialClusters = new ArrayList<>();

		for (Iris initialData : data.subList(0, k)) {
			KMeanCluster cluster = new KMeanCluster();
			cluster.ClusterMembers.add(initialData);
			initialClusters.add(cluster);
		}

		return initialClusters;
	}
}

File: Source/3 Data mining/Implementation - Jan/Dotnet/Itu.Jve.Datamining/DataminingConsole/Processes/DataMiningSpring2016/PatternDiscovery/Clustering/KMeans.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using DataminingConsole.Processes.DataMiningSpring2016.Entities;

namespace DataminingConsole.Processes.DataMiningSpring2016.PatternDiscovery.Clustering
{
    /// <summary>
    /// KMeans clusterings algorithm container.
    /// </summary>
    public class KMeans
    {
        /// <summary>
        /// Method to call to start the KMeansClustering process.
        /// </summary>
        /// <param name="k">number of clusters</param>
        /// <param name="data">dataset to cluster</param>
        /// <returns></returns>
        public static List<KMeanCluster> KMeansPartition(int k, List<DataTuple> data)
        {
            var clusters = GenerateInitialPartioningCluster(k, data);
            ReassignedResponse response;
            do
            {
                response = ReassignDataToCluster(data, clusters);
                clusters = response.Clusters;
            } while (response.Reassigned);

            return clusters;
        }

        /// <summary>
        /// Reassign tuples to nearest cluster
        /// </summary>
        /// <param name="data">dataset</param>
        /// <param name="clusters">currentclusters</param>
        /// <returns>new clusters + if it changed from currentclusters</returns>
        private static ReassignedResponse ReassignDataToCluster(IEnumerable<DataTuple> data, IEnumerable<KMeanCluster> clusters)
        {
            var newOldClusterIndex = new Dictionary<KMeanCluster, KMeanCluster>();

            var reassigned = false;

            // Create cluster index to map between current and new.
            foreach (var cluster in clusters)
            {
                var newCluster = new KMeanCluster { ClusterMean = cluster.CalculateClusterMean() };
                newOldClusterIndex.Add(newCluster, cluster);
            }

            // initialize new cluster
            var newClusters = new List<KMeanCluster>(newOldClusterIndex.Keys);

            // Assign tuples to new clusters.
            foreach (var record in data)
            {
                KMeanCluster minManhattenDistanceCluster = null;
                var minDistance = float.MaxValue;

                // Find closest cluster average.
                foreach (var cluster in newOldClusterIndex.Keys)
                {
                    var distance = Helpers.ManhattenDistance(record, cluster.ClusterMean);
                    
                    if (minDistance > distance)
                    {
                        minManhattenDistanceCluster = cluster;
                        minDistance = distance;
                    }
                }

                Debug.Assert(minManhattenDistanceCluster != null, "minManhattenDistanceCluster != null");

                minManhattenDistanceCluster.ClusterMembers.Add(record);

                // If changed cluster set reassigned to true.
                if (!reassigned && !newOldClusterIndex[minManhattenDistanceCluster].ClusterMembers.Contains(record))
                {
                    reassigned = true;
                }
            }

            return new ReassignedResponse { Clusters = newClusters, Reassigned = reassigned };
        }

        /// <summary>
        /// Use first k elements in data to create initial clusters
        /// </summary>
        /// <param name="k">number of clusters</param>
        /// <param name="data">dataset</param>
        /// <returns></returns>
        private static List<KMeanCluster> GenerateInitialPartioningCluster(int k, List<DataTuple> data)
        {
            if (data == null) throw new ArgumentNullException(nameof(data));

            return data.Take(k).Select(initialData => new KMeanCluster
            {
                ClusterMembers = new List<DataTuple> { initialData }
            }).ToList();
        }
    }
}
